File: src/visualization/interactivity/context_menus.py
# Copyright (c) 2025 MKM Research Labs. All rights reserved.
# 
# This software is provided under license by MKM Research Labs. 
# Use, reproduction, distribution, or modification of this code is subject to the 
# terms and conditions of the license agreement provided with this software.
# 
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

"""
Context menu functionality for interactive map markers.

This module handles the creation and management of right-click context menus
for property and gauge markers, extracted from the original visualization.py.
"""

from typing import Dict, Any, Optional, List
import folium
from pathlib import Path


class ContextMenuHandler:
    """Handler for context menu functionality on map markers."""
    
    def __init__(self):
        """Initialize the context menu handler."""
        self.property_menu_items = [
            {"id": "generate_report", "label": "📄 Generate PDF Report", "action": "generateReport"},
            {"id": "view_details", "label": "👁️ View Details", "action": "viewPropertyDetails"}
        ]
        
        self.gauge_menu_items = [
            {"id": "generate_gauge_report", "label": "📊 Generate Gauge Report", "action": "generateGaugeReport"},
            {"id": "view_time_series", "label": "📈 View Time Series", "action": "viewGaugeTimeSeries"},
            {"id": "export_data", "label": "💾 Export Data", "action": "exportGaugeData"}
        ]
    
    def get_property_context_menu_js(self) -> str:
        """
        Get JavaScript code for property context menus.
        
        Returns:
            JavaScript code as string
        """
        return self._load_js_template("property_context_menu.js")
    
    def get_gauge_context_menu_js(self) -> str:
        """
        Get JavaScript code for gauge context menus.
        
        Returns:
            JavaScript code as string
        """
        return self._load_js_template("gauge_context_menu.js")
    
    def get_base_context_menu_js(self) -> str:
        """
        Get base JavaScript code for context menu functionality.
        
        Returns:
            JavaScript code as string
        """
        js_code = f"""
        <script>
        // Global variables
        let currentPropertyId = null;
        let currentGaugeId = null;
        let mapInstance = null;
        
        // Store reference to the map when it's ready
        document.addEventListener('DOMContentLoaded', function() {{
            // Wait for the map to be fully loaded
            setTimeout(function() {{
                // Get the map instance from the global window object
                if (typeof window[Object.keys(window).find(key => key.startsWith('map_'))] !== 'undefined') {{
                    mapInstance = window[Object.keys(window).find(key => key.startsWith('map_'))];
                    console.log('Map instance found:', mapInstance);
                    initializeContextMenus();
                }} else {{
                    console.log('Map instance not found, retrying...');
                    setTimeout(arguments.callee, 500);
                }}
            }}, 1000);
        }});
        
        // Create context menu element
        function createContextMenu(menuId, menuItems) {{
            // Remove existing menu if it exists
            const existingMenu = document.getElementById(menuId);
            if (existingMenu) {{
                existingMenu.remove();
            }}
            
            const menu = document.createElement('div');
            menu.id = menuId;
            menu.style.cssText = 
                'position: absolute;' +
                'background: white;' +
                'border: 1px solid #ccc;' +
                'border-radius: 4px;' +
                'padding: 8px 0;' +
                'box-shadow: 0 2px 10px rgba(0,0,0,0.2);' +
                'z-index: 1000;' +
                'display: none;' +
                'font-family: Arial, sans-serif;' +
                'font-size: 14px;' +
                'min-width: 150px;';
            
            menuItems.forEach(item => {{
                const menuItem = document.createElement('div');
                menuItem.innerHTML = item.label;
                menuItem.style.cssText = 
                    'padding: 8px 16px;' +
                    'cursor: pointer;' +
                    'border-bottom: 1px solid #eee;';
                menuItem.onmouseover = () => menuItem.style.backgroundColor = '#f0f0f0';
                menuItem.onmouseout = () => menuItem.style.backgroundColor = 'white';
                menuItem.onclick = () => {{
                    if (item.action && typeof window[item.action] === 'function') {{
                        if (menuId.includes('property')) {{
                            window[item.action](currentPropertyId);
                        }} else if (menuId.includes('gauge')) {{
                            window[item.action](currentGaugeId);
                        }}
                    }}
                    hideContextMenu(menuId);
                }};
                
                menu.appendChild(menuItem);
            }});
            
            document.body.appendChild(menu);
            return menu;
        }}
        
        // Show context menu
        function showContextMenu(e, itemId, menuType) {{
            e.preventDefault();
            e.stopPropagation();
            
            if (menuType === 'property') {{
                currentPropertyId = itemId;
                showPropertyContextMenu(e);
            }} else if (menuType === 'gauge') {{
                currentGaugeId = itemId;
                showGaugeContextMenu(e);
            }}
            
            console.log('Context menu shown for ' + menuType + ': ' + itemId);
        }}
        
        // Show property context menu
        function showPropertyContextMenu(e) {{
            const menuItems = {self.property_menu_items};
            let menu = document.getElementById('property-context-menu');
            if (!menu) {{
                menu = createContextMenu('property-context-menu', menuItems);
            }}
            
            menu.style.left = e.pageX + 'px';
            menu.style.top = e.pageY + 'px';
            menu.style.display = 'block';
        }}
        
        // Show gauge context menu
        function showGaugeContextMenu(e) {{
            const menuItems = {self.gauge_menu_items};
            let menu = document.getElementById('gauge-context-menu');
            if (!menu) {{
                menu = createContextMenu('gauge-context-menu', menuItems);
            }}
            
            menu.style.left = e.pageX + 'px';
            menu.style.top = e.pageY + 'px';
            menu.style.display = 'block';
        }}
        
        // Hide context menu
        function hideContextMenu(menuId) {{
            const menu = document.getElementById(menuId);
            if (menu) {{
                menu.style.display = 'none';
            }}
        }}
        
        // Hide all context menus when clicking elsewhere
        document.addEventListener('click', function() {{
            hideContextMenu('property-context-menu');
            hideContextMenu('gauge-context-menu');
        }});
        
        // Function to extract property ID from marker tooltip
        function extractPropertyIdFromTooltip(tooltipText) {{
            if (!tooltipText) return null;
            
            // Look for "Property: XXXXX" pattern
            const match = tooltipText.match(/Property:\\s*([^|]+)/);
            if (match) {{
                return match[1].trim();
            }}
            return null;
        }}
        
        // Function to extract gauge ID from marker tooltip
        function extractGaugeIdFromTooltip(tooltipText) {{
            if (!tooltipText) return null;
            
            // FIRST: Look for specific GAUGE-XXXXX pattern (most precise)
            let match = tooltipText.match(/(GAUGE-[a-f0-9]+)/);
            if (match) {{
                return match[1].trim();
            }}
            
            // FALLBACK: Look for "Gauge: XXXXX" pattern (less precise)
            match = tooltipText.match(/Gauge:\s*([^|]+)/);
            if (match) {{
                return match[1].trim();  
            }}
            
            return null;
        }}
        
        // Function to extract ID from popup content
        function extractIdFromPopup(popupContent) {{
            if (!popupContent) return null;
            
            // Convert to string if it's not already
            let contentString = '';
            if (typeof popupContent === 'string') {{
                contentString = popupContent;
            }} else if (popupContent && typeof popupContent.toString === 'function') {{
                contentString = popupContent.toString();
            }} else if (popupContent && popupContent.innerHTML) {{
                contentString = popupContent.innerHTML;
            }} else if (popupContent && popupContent.textContent) {{
                contentString = popupContent.textContent;
            }} else {{
                console.log('Could not convert popup content to string:', typeof popupContent);
                return null;
            }}
            
            // Look for "ID: XXXXX" pattern in the popup HTML
            const match = contentString.match(/ID:\\s*([^<\\r\\n]+)/);
            if (match) {{
                return match[1].trim();
            }}
            return null;
        }}
        
        // Initialize context menu functionality
        function initializeContextMenus() {{
            if (!mapInstance) {{
                console.log('Map instance not available for context menu initialization');
                return;
            }}
            
            console.log('Initializing context menus...');
            
            // Function to add context menu to markers
            function addContextMenuToMarkers() {{
                let propertyMarkerCount = 0;
                let gaugeMarkerCount = 0;
                
                mapInstance.eachLayer(function(layer) {{
                    if (layer instanceof L.Marker) {{
                        // Skip if already has context menu listener
                        if (layer._hasContextMenu) return;
                        
                        // Try to determine marker type and extract ID
                        let markerId = null;
                        let markerType = null;
                        
                        // Method 1: From tooltip
                        if (layer.getTooltip && layer.getTooltip()) {{
                            const tooltipContent = layer.getTooltip().getContent();
                            
                            // Check for property marker
                            const propertyId = extractPropertyIdFromTooltip(tooltipContent);
                            if (propertyId) {{
                                markerId = propertyId;
                                markerType = 'property';
                            }} else {{
                                // Check for gauge marker
                                const gaugeId = extractGaugeIdFromTooltip(tooltipContent);
                                if (gaugeId) {{
                                    markerId = gaugeId;
                                    markerType = 'gauge';
                                }}
                            }}
                        }}
                        
                        // Method 2: From popup content
                        if (!markerId && layer.getPopup && layer.getPopup()) {{
                            const popupContent = layer.getPopup().getContent();
                            const extractedId = extractIdFromPopup(popupContent);
                            
                            if (extractedId) {{
                                if (extractedId.startsWith('PROP-')) {{
                                    markerId = extractedId;
                                    markerType = 'property';
                                }} else if (extractedId.startsWith('GAUGE-')) {{
                                    markerId = extractedId;
                                    markerType = 'gauge';
                                }}
                            }}
                        }}
                        
                        // Add context menu if we identified the marker
                        if (markerId && markerType) {{
                            // Add right-click event listener
                            layer.on('contextmenu', function(e) {{
                                showContextMenu(e.originalEvent, markerId, markerType);
                            }});
                            
                            // Mark as having context menu
                            layer._hasContextMenu = true;
                            layer._markerId = markerId;
                            layer._markerType = markerType;
                            
                            if (markerType === 'property') {{
                                propertyMarkerCount++;
                            }} else if (markerType === 'gauge') {{
                                gaugeMarkerCount++;
                            }}
                            
                            console.log('Added context menu to ' + markerType + ': ' + markerId);
                        }}
                    }}
                }});
                
                console.log('Added context menus to ' + propertyMarkerCount + ' property markers and ' + gaugeMarkerCount + ' gauge markers');
            }}
            
            // Add context menus to existing markers
            addContextMenuToMarkers();
            
            // Re-add context menus when new layers are added
            mapInstance.on('layeradd', function(e) {{
                setTimeout(function() {{
                    addContextMenuToMarkers();
                }}, 100);
            }});
            
            console.log('Context menu initialization complete');
        }}
        </script>
        """
        
        return js_code
    
    def _load_js_template(self, template_name: str) -> str:
        """
        Load JavaScript template from file.
        
        Args:
            template_name: Name of the template file
            
        Returns:
            JavaScript code as string
        """
        # For now, return empty string as templates would be in separate files
        # In a full implementation, these would be loaded from template files
        return ""
    
    def add_to_map(self, folium_map: folium.Map) -> None:
        """
        Add context menu functionality to a Folium map.
        
        Args:
            folium_map: The Folium map to add context menus to
        """
        # Add the base context menu JavaScript
        base_js = self.get_base_context_menu_js()
        folium_map.get_root().html.add_child(folium.Element(base_js))
        
        print("✅ Context menu functionality added to map")
    
    def configure(self, property_menu_items: Optional[List[Dict[str, str]]] = None,
                 gauge_menu_items: Optional[List[Dict[str, str]]] = None) -> None:
        """
        Configure context menu items.
        
        Args:
            property_menu_items: Custom property menu items
            gauge_menu_items: Custom gauge menu items
        """
        if property_menu_items:
            self.property_menu_items = property_menu_items
        
        if gauge_menu_items:
            self.gauge_menu_items = gauge_menu_items
        
        print("✅ Context menu items configured")
    
    def get_statistics(self) -> Dict[str, Any]:
        """
        Get statistics about context menu configuration.
        
        Returns:
            Dictionary with context menu statistics
        """
        return {
            'property_menu_items': len(self.property_menu_items),
            'gauge_menu_items': len(self.gauge_menu_items),
            'total_menu_items': len(self.property_menu_items) + len(self.gauge_menu_items)
        }

File: src/static/js/context-menus.js
// Copyright (c) 2022-2026 MKM Research Labs. All rights reserved.
//
// This software is licensed by MKM Research Labs for non-commercial
// research and educational use only. Any commercial use, including
// but not limited to use in or for products or services offered for sale,
// internal business operations intended for commercial advantage, or
// research and development conducted for a commercial entity, is expressly
// prohibited unless separately authorized in writing by MKM Research Labs.
//
// Use, reproduction, distribution, or modification of this code is subject to the
// terms and conditions of the license agreement provided with this software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// Context menu functionality for interactive map markers.

(function(root) {
    'use strict';

    function createContextMenuHandler(config) {
        config = config || {};
        var MENUS = {
            property: config.property || [],
            gauge: config.gauge || []
        };

        var currentId = null;
        var currentType = null;
        var currentDisplayName = null;
        var gaugesWithTrades = null; // null = not yet loaded; {} = loaded (may be empty)

        function createMenu(menuId, items, displayName) {
            var doc = root.document;
            var menu = doc.getElementById(menuId);
            if (menu) menu.remove();

            menu = doc.createElement('div');
            menu.id = menuId;
            menu.className = 'ctx-menu';

            // Header row with display name
            if (displayName) {
                var header = doc.createElement('div');
                header.className = 'ctx-menu-header';
                header.textContent = displayName;
                menu.appendChild(header);
            }

            items.forEach(function(item) {
                var el = doc.createElement('div');
                // Disable "Gauge Blotter" item when the current gauge has no open trades
                var isDisabled = item.id === 'gauge_blotter'
                    && gaugesWithTrades !== null
                    && !gaugesWithTrades[currentId];
                el.className = isDisabled
                    ? 'ctx-menu-item ctx-menu-item-disabled'
                    : 'ctx-menu-item';
                el.innerHTML = item.label;
                if (!isDisabled) {
                    el.onclick = function() {
                        if (item.action && root[item.action]) {
                            root[item.action](currentId, currentDisplayName);
                        }
                        hideMenu(menuId);
                    };
                }
                menu.appendChild(el);
            });

            doc.body.appendChild(menu);
            return menu;
        }

        function showMenu(e, id, type, displayName) {
            e.preventDefault();
            e.stopPropagation();

            currentId = id;
            currentType = type;
            currentDisplayName = displayName || null;

            var menuId = type + '-context-menu';
            var items = MENUS[type] || [];

            // Always recreate so header updates with current name
            var menu = createMenu(menuId, items, displayName);

            menu.style.left = e.pageX + 'px';
            menu.style.top = e.pageY + 'px';
            menu.style.display = 'block';
        }

        function hideMenu(menuId) {
            var menu = root.document.getElementById(menuId);
            if (menu) menu.style.display = 'none';
        }

        function hideAllMenus() {
            hideMenu('property-context-menu');
            hideMenu('gauge-context-menu');
            hideMenu('commercial-context-menu');
        }

        // Hide all menus on document click
        root.document.addEventListener('click', hideAllMenus);

        function extractId(content, pattern) {
            if (!content) return null;
            var str = typeof content === 'string' ? content :
                     (content.innerHTML || content.textContent || content.toString());
            var match = str.match(pattern);
            return match ? match[1].trim() : null;
        }

        function extractShortName(content, type) {
            // Extract a short display name from marker tooltip/popup content.
            // The shape of the tooltip differs per asset class, so we branch
            // on the resolved type — passing the wrong patterns at a tooltip
            // can otherwise capture stray hex chars and present them as
            // titles.
            if (!content) return null;
            var str = typeof content === 'string' ? content :
                     (content.innerHTML || content.textContent || content.toString());
            // Strip HTML tags
            str = str.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim();

            if (type === 'commercial') {
                // Commercial tooltip format:
                //   "{commercial_name} ({CommercialType})[ | Loan]"
                // We surface the name part as the menu title — fall back to
                // the CommercialType if no name is present.
                var commMatch = str.match(/^(.+?)\s*\(([^)]+)\)/);
                if (commMatch) {
                    var name = commMatch[1].trim();
                    if (name && !name.startsWith('CPROP-')) return name;
                    return commMatch[2].trim();  // fall back to CommercialType
                }
                return null;
            }

            // Property tooltip format: "Property: PROP-xxxxxxxx | Floods: ..."
            var propMatch = str.match(/Property:\s*(PROP-[a-f0-9]+)/);
            if (propMatch) return propMatch[1];

            // Gauge tooltip: text before "GAUGE-xxx" (Thames-prefixed area
            // names with _Point_N suffix).
            var gaugeMatch = str.match(/([A-Za-z][A-Za-z\s_]+?)(?:\s*[\|,]|\s+GAUGE-)/);
            if (gaugeMatch) {
                var gname = gaugeMatch[1].trim();
                var pi = gname.indexOf('_Point');
                if (pi > 0) gname = gname.substring(0, pi);
                gname = gname.replace(/^Thames\s+/i, '');
                if (gname.length > 0) return gname;
            }
            return null;
        }

        function initializeMenus() {
            var mapKey = Object.keys(root).find(function(k) { return k.startsWith('map_'); });
            var map = mapKey ? root[mapKey] : (typeof root.mapInstance !== 'undefined' ? root.mapInstance : null);

            if (!map) {
                console.log('Map not ready, retrying...');
                setTimeout(initializeMenus, 500);
                return;
            }

            var propCount = 0, gaugeCount = 0;

            function addMenusToMarkers() {
                map.eachLayer(function(layer) {
                    if (!(layer instanceof L.Marker) || layer._hasContextMenu) return;

                    var tooltip = layer.getTooltip && layer.getTooltip();
                    var popup = layer.getPopup && layer.getPopup();
                    var content = (tooltip && tooltip.getContent()) || (popup && popup.getContent()) || '';

                    var id = null, type = null;

                    // Order matters: CPROP- ids would substring-match the
                    // PROP- regex, so check commercial first.
                    id = extractId(content, /(CPROP-[a-f0-9]+)/);
                    if (id) type = 'commercial';

                    if (!id) {
                        id = extractId(content, /Property:\s*([^|<]+)/) || extractId(content, /(PROP-[a-f0-9]+)/);
                        if (id) type = 'property';
                    }

                    if (!id) {
                        id = extractId(content, /(GAUGE-[a-f0-9]+)/) || extractId(content, /Gauge:\s*([^|<]+)/);
                        if (id) type = 'gauge';
                    }

                    if (id && type) {
                        var name = extractShortName(content, type);
                        (function(boundId, boundType, boundName) {
                            layer.on('contextmenu', function(e) { showMenu(e.originalEvent, boundId, boundType, boundName); });
                        })(id, type, name);
                        layer._hasContextMenu = true;
                        layer._markerId = id;
                        layer._markerType = type;

                        if (type === 'property') propCount++;
                        else if (type === 'commercial') propCount++;
                        else gaugeCount++;
                    }
                });
            }

            addMenusToMarkers();
            map.on('layeradd', function() { setTimeout(addMenusToMarkers, 100); });

            // Pre-fetch which gauges have open trades so we can disable empty blotter items
            if (root.fetch) {
                root.fetch('/api/v1/trading/blotter/active-gauges')
                    .then(function(r) { return r.ok ? r.json() : Promise.reject(r.status); })
                    .then(function(data) {
                        gaugesWithTrades = {};
                        var ids = (data && data.gauge_ids) ? data.gauge_ids : [];
                        ids.forEach(function(gid) { gaugesWithTrades[gid] = true; });
                    })
                    .catch(function() {
                        // On any error, leave gaugesWithTrades = null so all items stay enabled
                        gaugesWithTrades = null;
                    });
            }

            console.log('Context menus: ' + propCount + ' properties, ' + gaugeCount + ' gauges');
        }

        function getCurrentId() { return currentId; }
        function getCurrentType() { return currentType; }

        function setActiveGauges(gaugeIds) {
            gaugesWithTrades = {};
            (gaugeIds || []).forEach(function(gid) { gaugesWithTrades[gid] = true; });
        }

        return {
            createMenu: createMenu,
            showMenu: showMenu,
            hideMenu: hideMenu,
            hideAllMenus: hideAllMenus,
            extractId: extractId,
            extractShortName: extractShortName,
            initializeMenus: initializeMenus,
            getCurrentId: getCurrentId,
            getCurrentType: getCurrentType,
            setActiveGauges: setActiveGauges
        };
    }

    // Node.js / Jest export
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = { createContextMenuHandler: createContextMenuHandler };
    }

    // Browser auto-init
    if (typeof root !== 'undefined' && root.document && root.__MENU_CONFIG) {
        var api = createContextMenuHandler(root.__MENU_CONFIG);

        if (root.document.readyState === 'loading') {
            root.document.addEventListener('DOMContentLoaded', function() { setTimeout(api.initializeMenus, 1000); });
        } else {
            setTimeout(api.initializeMenus, 1000);
        }
        root.setActiveGauges = api.setActiveGauges;
        if (root.console) root.console.log('Context menus initialized');
    }

})(typeof window !== 'undefined' ? window : (typeof global !== 'undefined' ? global : this));
